src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs

```diff
@@ -431,13 +451,26 @@ protected virtual void ReportUnassignedOutParameter(ParameterSymbol parameter, S
                             if (fieldSlot == -1 || !this.State.IsAssigned(fieldSlot))
                             {
                                 Symbol associatedPropertyOrEvent = field.AssociatedSymbol;
-                                if (associatedPropertyOrEvent?.Kind == SymbolKind.Property)
+                                bool hasAssociatedProperty = associatedPropertyOrEvent?.Kind == SymbolKind.Property;
+                                if (compilation.IsFeatureEnabled(MessageID.IDS_FeatureAutoDefaultStructs))
                                 {
-                                    Diagnostics.Add(ErrorCode.ERR_UnassignedThisAutoProperty, location, associatedPropertyOrEvent);
+                                    Diagnostics.Add(
+                                        hasAssociatedProperty ? ErrorCode.WRN_UnassignedThisAutoPropertySupportedVersion : ErrorCode.WRN_UnassignedThisSupportedVersion,
+                                        location,
+                                        hasAssociatedProperty ? associatedPropertyOrEvent : field);
                                 }
                                 else
                                 {
-                                    Diagnostics.Add(ErrorCode.ERR_UnassignedThis, location, field);
+                                    Diagnostics.Add(
+                                        hasAssociatedProperty ? ErrorCode.ERR_UnassignedThisAutoPropertyUnsupportedVersion : ErrorCode.ERR_UnassignedThisUnsupportedVersion,
+                                        location,
+                                        hasAssociatedProperty ? associatedPropertyOrEvent : field,
+                                        new CSharpRequiredLanguageVersion(MessageID.IDS_FeatureAutoDefaultStructs.RequiredVersion()));
+                                }
+
+                                if (TrackImplicitlyInitializedFields)
+                                {
+                                    this.AddImplicitlyInitializedField(field);
```

Comment:
nit: From offline discussion, the `if (TrackImplicitlyInitializedFields)` condition here is an optimization (we don't need to initialize hash set when we know the result will be discarded). For readability, I'm wonder if it would be better to move the `if (TrackImplicitlyInitializedFields)` inside `AddImplicitlyInitializedField`. Then it is clear that a diagnostic of some form is reported for all the same situations where auto-default feature kicks in.

Reply:
We could bake-in the `if (TrackImplicitlyInitializedFields)` to `AddImplicitlyInitializedField`, but it will be redundant with the check we do on line 1192, which we do to avoid the work of obtaining and enumerating the instance fields of 'this'. `TrackImplicitlyInitializedFields` is probably a very cheap call, so perhaps it is fine.

`AddImplicitlyInitializedField` currently does `Debug.Assert(TrackImplicitlyInitializedFields)`, FWIW. Let me know if you feel this should just change to `if (TrackImplicitlyInitializedFields)`.

Reply:
I'd lean towards baking it in (even if it means a redundant call), but I don't feel strongly. Maybe @cston can add his thought. This is up to you otherwise.
This is a nit, since from our discussion I'm convinced of correctness.